File: algorithms-well-known/_interval_search_tree.py
# https://www.youtube.com/watch?v=3kpmHEJXrsA
#
# --END--


class Node:

    def __init__(self, lo, hi):
        self.lo = lo
        self.hi = hi
        self.max = hi
        self.left = None
        self.right = None

    def __repr__(self):
        return '[{}, {}]'.format(self.lo, self.hi)


class IntervalTree:

    def __init__(self):
        self.root = None

    def insert(self, lo, hi):
        self.root = self._insert(self.root, lo, hi)

    def _insert(self, node, lo, hi):
        # when doing insertion, we use `lo` as key to do the insertions.
        if node is None:
            return Node(lo, hi)

        # go to left when less equal
        if lo <= node.lo:
            node.left = self._insert(node.left, lo, hi)
        else:
            node.right = self._insert(node.right, lo, hi)

        # update `max`
        node.max = max(node.max, hi)

        return node

    def search(self, lo, hi):
        """
        Find all possible intersections.
        """
        res = []
        self._search(self.root, lo, hi, res)
        return res

    def _search(self, node, lo, hi, res):
        # intersection is impossible
        if node is None or node.max < lo:
            return []

        # check current node
        if self._intersect(node, Node(lo, hi)):
            res.append(node)

        # check if right tree is possible
        if node.lo < hi:
            self._search(node.right, lo, hi, res)

        self._search(node.left, lo, hi, res)

    def _intersect(self, n1, n2):
        return n1.lo < n2.hi and n2.lo < n1.hi


# Example:
#
#                 [17, 19]
#               /          \
#              /            \
#        [5, 8]              [21, 24]
#       /     \
#      /       \
# [4, 8]       [15, 18]
#              /      \
#             /        \
#        [7, 10]       [16, 22]
#
#
# Search [21, 23]




tree = IntervalTree()
tree.insert(17, 19)
tree.insert(5, 8)
tree.insert(4, 8)
tree.insert(15, 18)
tree.insert(7, 10)
tree.insert(16, 22)
tree.insert(21, 24)
print(tree.search(21, 23))

File: algorithms-well-known/interval_search_tree.go
package main

import "fmt"

type Node struct {
    Lo int
    Hi int
    Max int
    Left *Node
    Right *Node
}

type IntervalTree struct {
    root *Node
}

func NewIntervalTree() IntervalTree {
    it := IntervalTree{root:nil}
    it.root = nil
    return it
}


func (it *IntervalTree) Insert(lo int, hi int) {
    it.root = it.insert(it.root, lo, hi)
}

func (it *IntervalTree) insert(node *Node, lo int, hi int) *Node {
    if node == nil {
        new_node := Node{Lo:lo, Hi:hi, Max:hi, Left:nil, Right:nil}
        return &new_node
    }
    if lo <= node.Lo {
        node.Left = it.insert(node.Left, lo, hi)
    } else {
        node.Right = it.insert(node.Right, lo, hi)
    }

    if node.Max < hi {node.Max = hi}
    return node
}

func (it *IntervalTree) Search(lo int, hi int) []Node {
    res := make([]Node, 0)
    it.search(it.root, lo, hi, &res)
    return res
}

func (it *IntervalTree) search(node *Node, lo int, hi int, res *[]Node) {
    if node == nil || node.Max < lo {
        return
    }

    if isIntersection(node, &Node{Lo:lo, Hi:hi}) {
        *res = append(*res, *node)
    }

    if node.Lo < hi {
        it.search(node.Right, lo, hi, res)
    }

    it.search(node.Left, lo, hi, res)
}

func isIntersection(n1 *Node, n2 *Node) bool {
    return n1.Lo < n2.Hi && n2.Lo < n1.Hi
}

// Testing Example:
//
//                 [17, 19]
//               /          \
//              /            \
//        [5, 8]              [21, 24]
//       /     \
//      /       \
// [4, 8]       [15, 18]
//              /      \
//             /        \
//        [7, 10]       [16, 22]
//
//
// Search [21, 23]

func main() {
    it := NewIntervalTree()
    it.Insert(17, 19)
    it.Insert(5, 8)
    it.Insert(4, 8)
    it.Insert(15, 18)
    it.Insert(7, 10)
    it.Insert(16, 22)
    it.Insert(21, 24)
    fmt.Print(it.Search(21, 23))
}


